utils/interface.py

```diff
@@ -0,0 +1,256 @@
+            self.date_created.replace(":", "-"),
+            bonjour.lower())
+        self.calibration_path = join(self.path, self.calibration_file)
+
+        self.messages_file = "{}_{}_messages.log".format(
+            self.date_created.replace(":", "-"),
+            bonjour.lower())
+        self.messages_path = join(self.path, self.messages_file)
+
+        if not isdir(self.path):
+            mkdir(self.path)
+
+        self.serial = SerialWrapper(baudrate=baudrate, bonjour=bonjour)
+
+    def write_data(self, dataArray):
+        """ Append a line in the file located at `self.data_path`
+
+        Each element of `dataArray` is written separated by a comma ','
+        If the file does not exit, it is created
+
+        Parameters
+        ----------
+        dataArray: array-like object
+            data to write in the file
+
+        """
+        with open(self.data_path, 'a+', newline='') as file:
+            writer = csv.writer(file, delimiter=',')
+            writer.writerow(dataArray)
+
+    def process_header(self, line):
+        """ Save header in memory
+
+        The first item is the computer time of reception
+
+        Parameters
+        ----------
+        line : string
+            line received from telemetry, without start and end character
+
+        """
+        # Don't write the header twice
+        if not self.header:
+            self.header = ["Computer Time"] + line.split(self.separators['SEP_DATA'])
+            self.write_data(self.header)
+            print("Header : {}".format(self.header))
+
+    def process_data(self, now, line):
+        """ Save data in memory and on file storage
+
+        The first item is the computer time of reception as a datetime.datetime object
+        Data cannot be recorded if the header has not been received
+
+        Parameters
+        ----------
+        now : datetime object
+            time of reception of the line
+        line : string
+            line received from telemetry, without start and end character
+
+        """
+        if self.header:
+            data = [now] + line.split(self.separators['SEP_DATA'])
+            self.data.append(data)
+            self.write_data(data)
+
+    def write_calibration(self):
+        """ Write the calibration data in the file located at `self.calibration_path`
+
+        """
+        with open(self.calibration_path, 'w') as file:
+            file.write(yaml.dump(self.calibration))
+
+    def process_calibration(self, line):
+        """ Save calibration data in memory and on file storage
+
+        Parameters
+        ----------
+        line : string
+            line received from telemetry, without start and end character
+
+        """
+        # Don't save the calibration twice
+        if not self.calibration:
+            self.calibration = {value.split(self.separators['SEP_CALI'])[0]: value.split(
+                self.separators['SEP_CALI'])[1] for value in line.split(self.separators['SEP_DATA'])}
```

Comment:
How is this working? Can you explain the syntax/ logic for this? As i understand it the purpose is to split up calibration values to a list but i don't get how the code does it.

Reply:
See #4 
The calibration data is sent as "key:value" strings separated by a "&". This line iterates over all strings between "&" characters (‘SEP_DATA’) in the line and then put the content as a key/value pair in a dictionary. Key and value are respectively before and after the ":" characters  (‘SEP_CALI’)